aspnetcore/mvc/advanced/custom-model-binding.md

```diff
@@ -0,0 +1,108 @@
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Controllers/ImageController.cs?name=post2&highlight=2)]
+
+As long as the binder can bind request data to appropriately named properties or arguments, model binding will succeed.
+
+## Implementing a custom model binder
+
+You can implement your own custom model binder that:
+
+- Converts incoming request data into properly-typed keys.
+- Uses Entity Framework Core to fetch the associated entity.
+- Passes the associated entity as an argument to the action method.
+
+The simplest approach is to use the `ModelBinderAttribute`. With this attribute, you specify the name of the input that is bound and the type of the `IModelBinderProvider` that it should use. The `ModelBinderAttribute` is applied to the type that should use the model binder:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Data/Author.cs?highlight=10)]
+
+The `AuthorEntityBinder` takes in an instance of `AppDbContext`. After confirming the expected input ("authorId" by default, or set using another `ModelBinderAttribute`) is present, it casts it to an integer and uses it to retrieve the associated entity. If successful, the result is set on the binding context. If model binding fails due to the input being the wrong type, a model state error is added. Otherwise, if no record is found for the id specified, the argument passed to the action method is set to `null`.
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Binders/AuthorEntityBinder.cs?name=demo)]
+
+The `AuthorEntityBinder` allows action methods to use the `Author` entity directly:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Controllers/BoundAuthorsController.cs?name=demo1&highlight=2)]
+
+In this example, since the name of the argument is not the default `authorId`, it's specified on the parameter using `ModelBinderAttribute`. Note that both the controller and action method are simplified compared to looking up the entity in the action method, since the `AppDbContext` is no longer required by the action or controller.
+
+You can apply the `ModelBinderAttribute` to individual model properties (such as on a viewmodel) or to action method parameters to specify a certain model binder or model name for just that type or action.
+
+### Implementing a ModelBinderProvider
+
+Instead of applying an attribute, you can implement [IModelBinderProvider](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.modelbinding.imodelbinderprovider). This is how the built-in framework binders are implemented. Note that when you specify the type your binder operates on, you specify the type of argument it produces, not the input your binder accepts. The following binder provider would work for the `AuthorEntityBinder` implemented above, and when added to MVC's collection of providers, would eliminate the need for the `ModelBinderAttribute` on `Author`.
```

Comment:
The following binder provider works with the `AuthorEntityBinder`. When it's added to MVC's collection of providers, you don't need to use the  `ModelBinder` Attribute on `Author` models.